solutions/gold/coci-17-sirni.mdx

```diff
@@ -0,0 +1,135 @@
+---
+id: coci-17-sirni
+source: COCI 2017
+title: Sirni
+author: Kevin Sheng
+---
+
+<Spoiler title="Hint 1">
+
+With $10^5$ cards, there's way too many possible edges brute force would have to consider.
+Is there any way we can cut out a majority of these possible edges?
+
+</Spoiler>
+
+<Spoiler title="Hint 2">
+
+Say you have a card with value $5$, and you have two other cards with values $12$ and $13$.
+Is it *ever* optimal to link $5$ to $13$?
+
+</Spoiler>
+
+<Spoiler title="Solution">
+
+## Explanation
+
+Following the first hint, let's try to build up a candidate list of edges that have some semblance of a chance at being included in Daniel's MST.
+
+First, let's sort the cards.
+We can also just not consider any duplicate cards, as those can be linked to other cards of the same value at zero cost.
+
+Now for each card $c$, we want to find all possible cards that might link up with it in an MST.
+Since edges go both ways, we only have to consider cards with values greater than $c$.
+
+Here's where the second hint comes in.
+No matter what other cards are in this example, it will always be optimal for $5$ to link up with $12$ instead of $13$.
+
+To generalize this, say we have a card with value $n$, and we have some values in between $kn$ and $(k+1) \cdot n$, where $k$ is some random multiple.
+Among these values, it will always make sense for us to choose the one closest to $kn$, as that will give the least cost.
```

Comment:
Probably a good idea to explain why this runs in time? I think bc of harmonic series right

Reply:
could probably prove time comp but idt it adds too much
if you really want to i'll slog through the process but

Reply:
you could just leave a note bc i was a bit confused as to how it was n log n at first and then i realized "oh harmonic series"

but to someone who's never heard of it they might be confused so maybe just put a wikipedia link or something?

Reply:
the n log n is bc of the sorting lmao

Reply:
harmonic series too no? when you loop through all the multiples

Reply:
perhaps

Reply:
bump? maybe just say something like "looping over all multiples of each number may seem too slow at first, but since they're all distinct, the total # of iterations is bounded by n/1 + n/2 + n/3...n/n, which is bounded by n log n" and then drop a link to harmonic series wikipedia or something

Reply:
yeah sure, just wait a lil

Reply:
lgtm